### ISSUE TITLE
FYI: upgrade to cosmwasm-typescript-gen makes binary export unneeded

### ISSUE THREAD

**Issue by pyramation** [nymlab/vectis#67]
see this PR for context: https://github.com/public-awesome/stargaze-contracts/pull/304/files

previously you had to manually export Binary, now it's taken care of by the transpiler.

```
npm install -g cosmwasm-typescript-gen@0.3.1
```

**Comment by pyramation** [comment]
@giovanni-orciuolo can you test out the latest and make sure all is ok?

I did a minor version bump just to mark a "breaking change" in case things don't work out and I need to fix anything on my end.

**Comment by j0nl1** [comment]
@pyramation I tried with the new version and now it generates the schema without binary.json which seems okay but after that got an error exporting RelayTx which use [RelayTransaction Structure](https://github.com/nymlab/vectis/blob/main/packages/vectis/src/wallet.rs#L33) which those fields are fields are of binary type.

```
Error: cannot find type for {"oneOf":[{"type":"object","required":["RelayTx"],"properties":{"RelayTx":{"type":"object","required":["message","nonce","signature","user_pubkey"],"properties":{"message":{"description":"Message to verify","allOf":[{"description":"Binary is a wrapper around Vec<u8> to add base64 de/serialization with serde. It also adds some helper methods to help encode inline.\n\nThis is only needed as serde-json-{core,wasm} has a horrible encoding for Vec<u8>","type":"string","title":"Binary"}]},"nonce":{"description":"Nonce","type":"integer","format":"uint64","minimum":0},"signature":{"description":"Serialized signature (message + nonce). Cosmos format (64 bytes). Cosmos format (secp256k1 verification scheme).","allOf":[{"description":"Binary is a wrapper around Vec<u8> to add base64 de/serialization with serde. It also adds some helper methods to help encode inline.\n\nThis is only needed as serde-json-{core,wasm} has a horrible encoding for Vec<u8>","type":"string","title":"Binary"}]},"user_pubkey":{"description":"User pubkey","allOf":[{"description":"Binary is a wrapper around Vec<u8> to add base64 de/serialization with serde. It also adds some helper methods to help encode inline.\n\nThis is only needed as serde-json-{core,wasm} has a horrible encoding for Vec<u8>","type":"string","title":"Binary"}]}},"title":"RelayTransaction","additionalProperties":true}},"additionalProperties":false},{"type":"object","required":["DirectMigrationMsg"],"properties":{"DirectMigrationMsg":{"description":"Binary is a wrapper around Vec<u8> to add base64 de/serialization with serde. It also adds some helper methods to help encode inline.\n\nThis is only needed as serde-json-{core,wasm} has a horrible encoding for Vec<u8>","type":"string","title":"Binary"}},"additionalProperties":false}],"title":"ProxyMigrationTxMsg"}
```

is there anything else that should be updated?

**Comment by pyramation** [comment]
thanks man! I'll look into this ASAP! tracking here https://github.com/pyramation/cosmwasm-typescript-gen/pull/22

**Comment by j0nl1** [comment]
thanks @pyramation really good job, I tested version 0.3.5 and is working pretty good!

I just notice that types has been removed from methods params, is there any motivation behind that?
![image](https://user-images.githubusercontent.com/40375118/175393340-a8cbeb41-c31f-4877-ae0e-cb25c750e390.png)

I think sometimes is easier for params as Addr because you don't have to worry about force Addr type to a string which is often used but in some complex structures is nice to have the interface.

**Comment by pyramation** [comment]
@j0nl1 this is not the intention... let me double check this one again. We want the actual objects, not `object`. Will work on a fix for this!

**Comment by j0nl1** [comment]
We should use `@cosmwasm/ts-codegen` npm package instead of `cosmwasm-typescript-gen` when there will be a solution for methods params.

Thanks to @sascha1337 for pointing that in the issue: https://github.com/nymlab/vectis/issues/73

**Comment by pyramation** [comment]
hey you guys! I got it fixed as you can see in our PR: 

```diff
-    newUser: string;
-    oldUser: string;
+    newUser: Addr;
+    oldUser: Addr;
```

and 

```diff
-    migrationMsg: object;
-    walletAddress: object;
+    migrationMsg: ProxyMigrationTxMsg;
+    walletAddress: WalletAddr;
```

To install globally:

(currently `0.7.1` but could be a later version by the time you read this)

```sh
npm install -g @cosmwasm/ts-codegen@0.7.1
```

or locally 

```sh
yarn add @cosmwasm/ts-codegen@0.7.1
```

**Comment by j0nl1** [comment]
@pyramation thanks for this update, amazing job! 🔝 
We finally could update to the latest version. 

I don't know if that is the expected behavoir but in proxy contract we are using a generic in the execute enum. We had an issue because the ts-generator created an alias `export type ProxyExecuteMsg = ExecuteMsg;` but ExecuteMsg was nothing because our execute msg was `ExecuteMsgForEmpty` for the generic.

I had to export with title to fix this [PR-82](https://github.com/nymlab/vectis/pull/82/files#diff-1613c519bd92465e5de979b01caa4ebf5f3a423ceda40f520ce80ed9d8146bfeR20)

**Comment by whalelephant** [comment]
@j0nl1 I am going to wait on @pyramation before merging #82  because the manual `export_with_title` is not intuitive for other dev as to why that was used.

**Comment by pyramation** [comment]
> @pyramation thanks for this update, amazing job! 🔝 We finally could update to the latest version.

Sure thing! happy to help!

> I don't know if that is the expected behavoir but in proxy contract we are using a generic in the execute enum. We had an issue because the ts-generator created an alias `export type ProxyExecuteMsg = ExecuteMsg;` but ExecuteMsg was nothing because our execute msg was `ExecuteMsgForEmpty` for the generic.

Hey this was added feature we can make it an optional flag and default it to false. Will fix asap.

**Comment by pyramation** [comment]
fixed https://github.com/CosmWasm/ts-codegen/commit/02fcbc4d88ccf8f3317216402ac33d31653e91f0

this removes the aliased `ExecuteMsg` and puts it under a flag that defaults to false.

```
Successfully published:
 - @cosmwasm/ts-codegen@0.7.4
 ```
 
 was there any other issue? or would this make it ready?

**Comment by j0nl1** [comment]
thanks again @pyramation this makes the pull request ready to be approved by @whalelephant

**Comment by whalelephant** [comment]
thank you so much @pyramation! great job!

**Comment by pyramation** [comment]
another decently big release on `@cosmwasm/ts-codegen` that can make the CI/CD part of this a LOT better :atom: :tada:

* programmatic usage + docs for CI/CD and batching multiple contracts

docs https://github.com/CosmWasm/ts-codegen#programmatic-usage

* contracts bundles — ability to generate bundles, handy for compiling libraries

docs https://github.com/CosmWasm/ts-codegen#bundles (not sure if this will be useful?)

POSSIBLE BREAKING CHANGES with `@cosmwasm/ts-codegen@0.9.2`

CLI: now you must specify `--plugin` for the `generate` command or you'll get an interactive prompt. Here is an example of how to specify one or more plugins using the CLI: https://github.com/CosmWasm/ts-codegen#specifying-plugins

HOWEVER PLEASE NOTE:

we now recommend for CI/CD using the programmatic usage, NOT the CLI... you will have a MUCH BETTER result :slight_smile: Happy to help with this, for now here are the docs https://github.com/CosmWasm/ts-codegen#programmatic-usage


```
@cosmwasm/ts-codegen@0.9.2
```

**Comment by pyramation** [comment]
sorry to iterate so fast on the version, I think this is the last one for a while 😅

I just wanted to make more of a plugin interface that will work better with CI/CD and start to move away from using the CLI and start using a more programmatic approach that is more sustainable with higher leverage!

**Comment by sascha1337** [comment]
> programmatic usage + docs for CI/CD and batching multiple contracts

Godlike update, this was some heavy blocker for me in multiple projects. Thanks for that one.

@pyramation do you think we can have the equivalent for cosmology repo? AFAIK, that one doesn't even allow cli args, forcing me into interactive cli mode regardless of the argv params- blocking issue for multiple repos I'm working on, too.

Would be fire 🔥 if that could get updated too

**Comment by pyramation** [comment]
> > programmatic usage + docs for CI/CD and batching multiple contracts
> 
> Godlike update, this was some heavy blocker for me in multiple projects. Thanks for that one.
> 
> @pyramation do you think we can have the equivalent for cosmology repo? AFAIK, that one doesn't even allow cli args, forcing me into interactive cli mode regardless of the argv params- blocking issue for multiple repos I'm working on, too.
> 
> Would be fire 🔥 if that could get updated too

@sascha1337 happy to help, pm me with the use-cases for cosmology and can show how I have that setup, I think I have it covered w the core

**Comment by sascha1337** [comment]
sure u can bypass interactive cli mode already? w8 lemme inspec'
